src/impl/tpls/KokkosBlas3_gemm_tpl_spec_decl.hpp

```diff
@@ -335,6 +335,127 @@ KOKKOSBLAS3_CGEMM_BLAS( Kokkos::LayoutRight, Kokkos::LayoutRight, Kokkos::Layout
 namespace Impl {
 
+
+// DotBasedGEMM implements the optimization for C = beta*C + alpha*A^TB 
+// with A and B matrices both being tall and skinny. C matrix is assumably 
+// small, so, each entry of C is computed by performing the dot product of 
+// respective columns of A and B matrices. Note that the dot products are
+// performed on very long vectors, so, each dot product is distributed among
+// numDivPerDot teams.     
+
+struct TagInit{};   // The tag for the initialization parallel_for 
+struct TagMult{};   // The tag for the multiplication parallel_for 
+template<class ExecSpace, class AV, class BV, class CV>
+struct DotBasedGEMM{
+
+  AV A;
+  BV B;
+  CV C;
+
+  typedef typename AV::non_const_value_type scalar_A;
+  typedef typename AV::size_type size_A;
+  typedef typename CV::non_const_value_type scalar_C;
+  typedef typename CV::size_type size_C;
+
+  scalar_A alpha;
+  scalar_C beta;
+
+  // The following types (especially dotSize) could have simply been int,
+  // e.g., see lines 488-490.
+  size_C numCrows;           
+  size_C numCcols;
+
+  size_C numDivPerDot;   // number of teams collectively performing a dot product
+  size_C numTeams;       // total number of teams
+  
+  size_A dotSize;        // the length of the vectors in the dot products
+  size_A chunkSize;      // the local length of each team's share on the dot product  
+  
+
+  DotBasedGEMM(const scalar_A& alpha_, const AV& A_, const BV& B_, const scalar_C& beta_, const CV& C_):A(A_),B(B_),C(C_),alpha(alpha_),beta(beta_) 
+  {
+    numCrows = C.extent(0);
+    numCcols = C.extent(1);
+    dotSize = A.extent(0);
+  }
+
+  void run() {
+
+    size_C ndots = C.extent(0) * C.extent(1);       // Number of dot products
+    size_C appxNumTeams = (dotSize * ndots) / 4096; // First, estimate the appxNumTeams
+                                                    // each team performs 4096 mult-add ops
+
+    // Adjust appxNumTeams in case it is too small or too large
+    if(appxNumTeams < 32)
+      appxNumTeams = 32;
+    if(appxNumTeams > 1024)
+      appxNumTeams = 1024;
+
+    // If there are more dot products than the number of teams,
+    // then set the number of teams to be number of dot products
+    // and each team will perform only one dot product.
+    // We don't want a team to perform more than one dot product.
+    if(ndots >= appxNumTeams) {
+      numTeams = ndots;
+      numDivPerDot = 1;
+    }
+    // If there are more teams than dot products, each dot product can
+    // potentially be performed by multiple teams. First, compute 
+    // numDivPerDot as an integer (take the floor, not ceiling), then,
+    // compute actual number of teams by using this factor.
+    else{
+      numDivPerDot = appxNumTeams / ndots;
+      numTeams = ndots * numDivPerDot;
+    }
+
+    // Determine the local length for the dot product
+    chunkSize = dotSize / numDivPerDot;
+    if(numDivPerDot > 1)
+      chunkSize++;
+
+    // Initialize C matrix as beta*C
```

Comment:
Add a comment to say this may not be needed because Belos does initialize, but we have to be careful

Reply:
Done.